capnp/src/private/arena.rs

```diff
@@ -31,21 +32,25 @@ use crate::{Error, OutputSegments, Result};
 pub type SegmentId = u32;
 
 pub struct ReadLimiter {
-    pub limit: Cell<u64>,
+    pub limit: u64,
+    pub read: AtomicU64,
 }
 
 impl ReadLimiter {
     pub fn new(limit: u64) -> ReadLimiter {
-        ReadLimiter { limit: Cell::new(limit) }
+        ReadLimiter { 
+            limit,
+            read: AtomicU64::new(0),
+        }
     }
 
     #[inline]
     pub fn can_read(&self, amount: u64) -> Result<()> {
-        let current = self.limit.get();
-        if amount > current {
+        let read = self.read.fetch_add(amount, Ordering::Relaxed) + amount;
```

Comment:
This potentially allows `self.read` to overflow and wrap around.

Reply:
I thought about it, but it also return an error at first if it reaches the limit, so I would expect this to be an undefined behaviour to call it afterward.

My first implementation was a load + cas, but that would be definitely less performant. 

We could also have a second AtomicBool that indicates that we reached the limit and return right away. That could potentially be over the limit, but decreases odds of overflow.